gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/matching/single/cypher/planning/planner/greedy/GreedyPlanner.java

```diff
@@ -163,7 +165,11 @@ private void createVertexPlans(PlanTable planTable) {
       CNF vertexPredicates = allPredicates.removeSubCNF(vertexVariable);
       Set<String> projectionKeys = allPredicates.getPropertyKeys(vertexVariable);
 
-      FilterAndProjectVerticesNode node = new FilterAndProjectVerticesNode(graph.getVertices(),
+      DataSet<org.gradoop.common.model.impl.pojo.Vertex> vertices =
+        vertex.getLabel().equals(GradoopConstants.DEFAULT_VERTEX_LABEL) ?
+          graph.getVertices() : graph.getVerticesByLabel(vertex.getLabel());
+
+      FilterAndProjectVerticesNode node = new FilterAndProjectVerticesNode(vertices,
```

Comment:
This could be further improved by checking if there are more predicates for this set of vertices to be evaluated. If not `ProjectVertices` can be used instead